target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-rax120v2.dts

```diff
@@ -0,0 +1,522 @@
+			color = <LED_COLOR_ID_GREEN>;
+		};
+
+		led_aqr_red {
+			label = "red:aqr";
+			gpios = <&led_gpio 10 GPIO_ACTIVE_LOW>;
+			color = <LED_COLOR_ID_RED>;
+		};
+
+		led_aqr_white {
+			label = "white:aqr";
+			gpios = <&led_gpio 11 GPIO_ACTIVE_LOW>;
+			color = <LED_COLOR_ID_WHITE>;
+		};
+
+		led_wps_white {
+			label = "white:wps";
+			gpios = <&tlmm 40 GPIO_ACTIVE_HIGH>;
+			color = <LED_COLOR_ID_WHITE>;
+		};
+	};
+};
+
+&clocks {
+	g762_clk: fixedclk {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <32768>;
+	};
+};
+
+&tlmm {
+	mdio_pins: mdio-pins {
+		mdc {
+			pins = "gpio68";
+			function = "mdc";
+			drive-strength = <8>;
+			bias-pull-up;
+		};
+
+		mdio {
+			pins = "gpio69";
+			function = "mdio";
+			drive-strength = <8>;
+			bias-pull-up;
+		};
+	};
+};
+
+&mdio {
+	status = "okay";
+
+	pinctrl-0 = <&mdio_pins>;
+	pinctrl-names = "default";
+	reset-gpios = <&tlmm 37 GPIO_ACTIVE_LOW>;
+
+	qca8075_0: ethernet-phy@0 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0>;
+	};
+
+	qca8075_1: ethernet-phy@1 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <1>;
+	};
+
+	qca8075_2: ethernet-phy@2 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <2>;
+	};
+
+	qca8075_3: ethernet-phy@3 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <3>;
+	};
+
+	qca8075_4: ethernet-phy@4 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <4>;
+	};
+
+	aqr111: ethernet-phy@7 {
+		compatible ="ethernet-phy-ieee802.3-c45";
+		reg = <7>;
+		reset-gpios = <&tlmm 59 GPIO_ACTIVE_LOW>;
+	};
+};
+
+&switch {
+	status = "okay";
+
+	switch_lan_bmp = <0x3e>; /* lan port bitmap */
+	switch_wan_bmp = <0x40>; /* wan port bitmap */
+	switch_mac_mode = <0x0>; /* mac mode for uniphy instance0*/
+	switch_mac_mode1 = <0xff>; /* mac mode for uniphy instance1*/
+	switch_mac_mode2 = <0xd>; /* mac mode for uniphy instance2*/
+	bm_tick_mode = <0>; /* bm tick mode */
+	tm_tick_mode = <0>; /* tm tick mode */
+
+	qcom,port_phyinfo {
+		port@1 {
+			port_id = <1>;
+			phy_address = <0>;
+		};
+		port@2 {
+			port_id = <2>;
+			phy_address = <1>;
+		};
+		port@3 {
+			port_id = <3>;
+			phy_address = <2>;
+		};
+		port@4 {
+			port_id = <4>;
+			phy_address = <3>;
+		};
+		port@5 {
+			port_id = <5>;
+			phy_address = <4>;
+		};
+		port@6 {
+			port_id = <6>;
+			phy_address = <7>;
+			compatible = "ethernet-phy-ieee802.3-c45";
+			ethernet-phy-ieee802.3-c45;
+		};
+	};
+};
+
+&edma {
+	status = "okay";
+};
+
+&dp1 {
+	status = "okay";
+	phy-handle = <&qca8075_0>;
+	label = "lan4";
+};
+
+&dp2 {
+	status = "okay";
+	phy-handle = <&qca8075_1>;
+	label = "lan3";
+};
+
+&dp3 {
+	status = "okay";
+	phy-handle = <&qca8075_2>;
+	label = "lan2";
+};
+
+&dp4 {
+	status = "okay";
+	phy-handle = <&qca8075_3>;
+	label = "lan1";
+};
+
+&dp5 {
+	status = "okay";
+	phy-handle = <&qca8075_4>;
+	label = "wan";
+};
+
+&dp6_syn {
+	status = "okay";
+	phy-handle = <&aqr111>;
+	label = "lan5";
+};
+
+
```

Comment:
Extra newline.

Also, how are the interface MAC-s being set?

Reply:
The follow was in @jewwest DTS and I did remove it (and added it again now). But only dp1 to dp4 are set correct.

```
 24         ethernet0 = &dp1;
 25         ethernet1 = &dp2;
 26         ethernet2 = &dp3;
 27         ethernet3 = &dp4;
 28         ethernet4 = &dp5;
 29         ethernet5 = &dp6_syn;
```

The MACs are stored in u-boot env variables:

```
eth1addr=6c:cd:d6:c0:7b:cd
eth2addr=6c:cd:d6:c0:7b:ce
eth3addr=70:9f:1f:77:27:28
eth4addr=35:30:30:32:32:34
eth5addr=35:38:36:4b:30:31
ethact=eth0
ethaddr=6c:cd:d6:c0:7b:cc
```

The relevant boot log is:

```
[    3.527317] ssdk_switch_device_num_init[1335]:INFO:ess-switch dts node number: 1
[    3.527422] ssdk_dt_get_switch_node[1124]:INFO:ess-switch DT exist!
[    3.533813] ssdk_dt_parse_access_mode[999]:INFO:switch_access_mode: local bus
[    3.539780] ssdk_dt_parse_access_mode[1012]:INFO:switchreg_base_addr: 0x3a000000
[    3.547075] ssdk_dt_parse_access_mode[1013]:INFO:switchreg_size: 0x1000000
[    3.554540] ssdk_dt_parse_mac_mode[327]:INFO:mac mode = 0x0
[    3.561216] ssdk_dt_parse_mac_mode[336]:INFO:mac mode1 = 0xff
[    3.566689] ssdk_dt_parse_mac_mode[345]:INFO:mac mode2 = 0xd
[    3.572634] ssdk_dt_parse_uniphy[364]:INFO:ess-uniphy DT exist!
[    3.578412] ssdk_dt_parse_intf_mac[942]:INFO:dp1 MAC 6c:cd:d6:c0:7b:cc
[    3.583995] ssdk_dt_parse_intf_mac[942]:INFO:dp2 MAC 6c:cd:d6:c0:7b:cd
[    3.590587] ssdk_dt_parse_intf_mac[942]:INFO:dp3 MAC 6c:cd:d6:c0:7b:ce
[    3.597102] ssdk_dt_parse_intf_mac[942]:INFO:dp4 MAC 70:9f:1f:77:27:28
...
[    5.071711] Generic PHY 90000.mdio-1:00: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:00, irq=POLL)
[    5.075836] nss-dp 3a001000.dp1 lan4: Registered netdev lan4(qcom-id:1)
[    5.085208] Generic PHY 90000.mdio-1:01: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:01, irq=POLL)
[    5.091960] nss-dp 3a001200.dp2 lan3: Registered netdev lan3(qcom-id:2)
[    5.101364] Generic PHY 90000.mdio-1:02: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:02, irq=POLL)
[    5.108074] nss-dp 3a001400.dp3 lan2: Registered netdev lan2(qcom-id:3)
[    5.117498] Generic PHY 90000.mdio-1:03: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:03, irq=POLL)
[    5.124191] nss-dp 3a001600.dp4 lan1: Registered netdev lan1(qcom-id:4)
[    5.133519] GMAC5(ffffff8002d7d8c0) Invalid MAC@ - using e2:13:3c:1a:59:eb
[    5.139919] Generic PHY 90000.mdio-1:04: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:04, irq=POLL)
[    5.147267] nss-dp 3a001800.dp5 wan: Registered netdev wan(qcom-id:5)
[    5.156540] GMAC6(ffffff8002e208c0) Invalid MAC@ - using ca:18:79:26:23:33
[    5.163085] Generic Clause 45 PHY 90000.mdio-1:07: attached PHY driver (mii_bus:phy_addr=90000.mdio-1:07, irq=POL)
[    5.170083] nss-dp 3a007000.dp6-syn lan5: Registered netdev lan5(qcom-id:6)
```

Reply:
If they are stored in U-boot env then those can be used via NVMEM to populate MAC-s, cause having random MAC-s doesn't really sound normal.

Reply:
Maybe it was that part of the [discussion from @jewwest PR](https://github.com/openwrt/openwrt/pull/11830#discussion_r1081879908)?

> @jewwest : I think we started chatting about this before. SMEM provides the partition layout represented by QCA however this is incomplete. Netgear overrides it by hard coding their partition layout in ofpart.c that was used as a foundation for these entries.

Can I still use NVRAM?

But I'll try again, I did try using NVMEM similar to how 'ipq8074-wxr-5950ax12' or 'ipq8074-nbg7815.dts' or 'ipq8072-wax218.dts use it but was not successful.

Reply:
Gave NVMEM a go and it didn't work. It seems someone is not happy about the MACs that start with '35:30:30' and '35:38:36'. According to MAC lookup they are not assigned (but that seems to be true for the '70:9f:1f' one, too).

All the MACs are saved in the 'boarddata1' partition, starting from offset 0. I replaced the two '35:...' MACs with valid one and then the interfaces get theses MACs successfully. And it works then with both methods, using ethernetX aliases and NVRAM.

So I don't know how to resolve that. How to generate unique MACs and have them assigned in another way?

Reply:
Those `35:..` MAC-s are probably invalid so they are rejected.

So, stock FW assigns random MAC-s on each boot?

Reply:
Yes, in stock firmware it does for me if I run `ethtool -P eth4` or `ethtool -P eth5`, after every reboot I get a new pair of MACs.

`arp` on my connecting client or uplink report the same MAC for all the LAN and for WAN it's "LAN's MAC + 1"

Reply:
But the rest are static?

If so, then at least those should be set, ideally using NVMEM.

Reply:
Yes the rest is static and they all are taken from the 'boardata1' partition.

NVMEM: Okey, but that adds a few lines of code and doesn't seem to behave different then when I set alias. The alias methods picked up the MAC addresses change in the boarddata1 partition. Is the aliases method not recommended?

E.g. I add:

```
 &dp1 {
        status = "okay";
        phy-handle = <&qca8075_0>;
        label = "lan4";
+       nvmem-cells = <&macaddr_dp1>;
+       nvmem-cell-names = "mac-address";
 };

 &dp2 {
        status = "okay";
        phy-handle = <&qca8075_1>;
        label = "lan3";
+       nvmem-cells = <&macaddr_dp2>;
+       nvmem-cell-names = "mac-address";
 };
[and so on ...]
                                label = "boarddata1";
                                reg = <0x1180000 0x0100000>;
                                read-only;
+
+                               compatible = "nvmem-cells";
+                               #address-cells = <1>;
+                               #size-cells = <1>;
+
+                               macaddr_dp1: macaddr@0 {
+                                       reg = <0x0 0x6>;
+                               };
+
+                               macaddr_dp2: macaddr@1 {
+                                       reg = <0x6 0x6>;
+                               };
[and so on ...]
```

That works but as said the result is the same as with aliases.

Reply:
And since the 'wan' interface is one of the two interfaces with invalid MAC: How can it made sure that the WAN interface gets a correct permanent MAC? Because I assume an ever changing MAC won't be going down well with PPPoE and other services.

Reply:
@robimarko do you have an idea about the changing MAC addresses, especially for the WAN interface?

Reply:
Well, if the stock FW does the same and there is no MAC allocated for those not much can be done, we cant invent a MAC for them, its a shitty move from the vendor honestly.

Reply:
I see; in stock although the two MACs are generated after every reboot; since they create a bridge for WAN too the MAC showing up in ARP is "LAN-base-MAC + 1". Not sure if that behavior could (or should) be replicated in OpenWrt.